### PR TITLE
Initial capability to create conferences and join them.

### PR DESCRIPTION
**Pull request by snicki13**
Available functionality:
Create classrooms and join via BBB API.
It is possible to configure the feedback system (or probably any BBB Gateway, like Greenlight or PILOS) to connect to the digital classroom instead of a BBB cluster.
Use http://localhost:8085 as address.

Tickets and conferences can be created within the classroom. Invites or automatic refreshing of tickets and users is WIP.

### CODE REVIEW

File: gateway-api.yaml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+info:
+  title: Digital Classroom Gateway API
+  description: API to control the digital classroom application from the gateway service. This APi is based on the Big Blue Button API.
+  version: 0.0.1
+
+tags:
+  - name: Create Room
+    description: Creating a new digital classroom
+servers:
+  - url: 'https://localhost:8080/api/v1'
```
Comment (mxsph) [inline gateway-api.yaml:11]:
```suggestion
  - url: 'https://localhost:8085/api/v1'
```

The README says to use `https://localhost:8085`, shouldn't that be here too?

Comment (snicki13) [reply gateway-api.yaml:11]: You‘re right. I just haven‘t updated this for a while. Thank you.

**Review by mxsph — CHANGES_REQUESTED** [review]
I noticed two errors during the test:

- When opening the `digital-classroom` from the `FBS`, the browser asks for a user name and a password (see `gif`). After some experimenting I found out that the two slashes in the url (http://localhost:8085//api/join?fullName=...) are the problem (this is a bug in the `FBS` but should not lead to a problem).
- If you remove the one slash it works so far, but you are not redirected to the conference (i.e. the welcome message), you just get an xml response. As far as I read in the [BBB specification](https://docs.bigbluebutton.org/dev/api.html#join), the `join` route should redirect to the conference by default and only give an `xml` response if you specify `redirect=FALSE`.

error

**Comment by snicki13** [comment]
The requested changes were adressed within commits. There are still many open ends, but it is advised to get something relatively stable to the master branch to do first integration tests with PILOS and the FBS.

**Review by mxsph — CHANGES_REQUESTED** [review]
> There are still many open ends, but it is advised to get something relatively stable to the master branch to do first integration tests with PILOS and the FBS.

:+1: 

I only noticed one more mistake that we should fix before merging. In the `Dockerfile` the `jar` is simply copied into the container. 

https://github.com/thm-mni-ii/digital-classroom/blob/bed214e2a1a88ff6e2e5659076a93ff75889f7d3/Dockerfile#L8

In order for this to work, you have to assemble them beforehand.
I always find this a little unattractive because you then have to have everything installed (furthermore, the current CI/CD will not work with it). I would therefore adapt the `Dockerfile` in such a way that it uses [ multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) and thus assembles the programme in the first stage and copies the `jar` from here.

**Comment by snicki13** [comment]
The artifact produced by cradle contains every dependency of the main project. If I understood your inquiry correctly, that was your main concern. 
So you are suggesting to build the artifact within the docker build process?

**Comment by mxsph** [comment]
> The artifact produced by Gradle contains every dependency of the main project. If I understood your inquiry correctly, that was your main concern.

No. My problem is that in order to build the Docker container, I would have to build the program `local` first and thus have `jdk` etc. installed. Furthermore, I would also need to run two commands to build the container.
 

> So you are suggesting to build the artifact within the docker build process?

Yes, but in order to avoid unnecessary programs (such as the build tools) in the final Docker container, you can use  [ multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/).

**Comment by snicki13** [comment]
> I noticed two errors during the test:
> 
>     * When opening the `digital-classroom` from the `FBS`, the browser asks for a user name and a password (see `gif`). After some experimenting I found out that the two slashes in the url (http://localhost:8085//api/join?fullName=...) are the problem (this is a bug in the `FBS` but should not lead to a problem).
> 
>     * If you remove the one slash it works so far, but you are not redirected to the conference (i.e. the welcome message), you just get an xml response. As far as I read in the [BBB specification](https://docs.bigbluebutton.org/dev/api.html#join), the `join` route should redirect to the conference by default and only give an `xml` response if you specify `redirect=FALSE`.
> 
> 
> error



> > There are still many open ends, but it is advised to get something relatively stable to the master branch to do first integration tests with PILOS and the FBS.
> 
> 👍
> 
> I only noticed one more mistake that we should fix before merging. In the `Dockerfile` the `jar` is simply copied into the container.
> 
> https://github.com/thm-mni-ii/digital-classroom/blob/bed214e2a1a88ff6e2e5659076a93ff75889f7d3/Dockerfile#L8
> 
> In order for this to work, you have to assemble them beforehand.
> I always find this a little unattractive because you then have to have everything installed (furthermore, the current CI/CD will not work with it). I would therefore adapt the `Dockerfile` in such a way that it uses [ multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/) and thus assembles the programme in the first stage and copies the `jar` from here.

fixed

**Review by mxsph — APPROVED** [review]

**Pull request merged**